File: 237G_Computational/Projects/Project3/Python/Question5.py
import math
import numpy as np
import matplotlib.pyplot as plot
import statistics
import numpy.linalg.cholesky as chol

def question5():
    seq1 = halton_seq(100, 2);
    seq2 = halton_seq(100, 7);

    plot.figure("Figure 1")
    plot.plot(seq1, seq2,"o")
    plot.show()

    seq3 = halton_seq(100, 4)
    plot.figure("Figure 2")
    plot.plot(seq1, seq3, "o")
    plot.show()

    ans_1 = calculate_integral(2, 4, 10000)
    ans_2 = calculate_integral(2, 7, 10000)
    ans_3 = calculate_integral(5, 7, 10000)

    mean_1 = statistics.mean(ans_1)
    mean_2 = statistics.mean(ans_2)
    mean_3 = statistics.mean(ans_3)
    print(mean_1,mean_2,mean_3)


def halton_seq(k,m):
    sequence = [0] * k
    r = math.ceil(math.log(k) / math.log(m)) + 1

    base_rep = [0] * r
    for count in range(0,k):
        j = 0
        done = False
        while not done:
            base_rep[j] = base_rep[j] + 1
            if base_rep[j] < m:
                done = True
            else:
                base_rep[j] = 0
                j = j+1

        for baseCt in range(0,r):
            sequence[count] = sequence[count] + (base_rep[baseCt] / m **(baseCt + 1))

    return sequence


def calculate_integral(m1, m2, k):
    halton_1 = np.array(halton_seq(k, m1))
    halton_2 = np.array(halton_seq(k, m2))

    return np.exp(- halton_1 * halton_2) * (np.sin(6 * math.pi * halton_1) +
                                            ((np.abs(np.cos(2 * math.pi * halton_2))**(1./3.))*
                                            [1 if math.cos(2*math.pi*val)>=0 else -1 for val in halton_2]))

File: 237G_Computational/Projects/Project3/Test/Project3/Project3/Question5.cpp
#define _USE_MATH_DEFINES
#include <cmath>
#include "Question5.h"
using namespace std;

vector<double> Question5(double seed)
{
	int N = 100;

	//a
	vector<double> uni_rand = GenerateUniformDist(N * 2, seed);
	vector<vector<double>> uni_vectors;
	int count = 0;
	for (int i = 0; i < N; i++) 
	{
		vector<double> vec;
		vec.push_back(uni_rand[count]);
		count = count + 1;
		vec.push_back(uni_rand[count]);
		count = count + 1;
		uni_vectors.push_back(vec);
	}
	writeToExcel("Excels/uniformDist.csv", uni_vectors);

	//b
	vector<double> seq1 = HaltonSequence(100, 2);
	vector<double> seq2 = HaltonSequence(100, 7);
	writeToExcel("Excels/HaltonSeq1.csv", seq1, seq2);
	//c
	vector<double> seq3 = HaltonSequence(100, 4);
	writeToExcel("Excels/HaltonSeq2.csv", seq1, seq3);

	//e
	int Tot = 10000;
	double answer1 = CalculateIntegral(2, 4, Tot);
	double answer2 = CalculateIntegral(2, 7, Tot);
	double answer3 = CalculateIntegral(5, 7, Tot);
	cout << endl << "Answer for Integral with bases 2 and 4 = " << answer1 << endl;
	cout << "Answer for Integral with bases 2 and 7 = " << answer2 << endl;
	cout << "Answer for Integral with bases 5 and 7 = " << answer3 << endl;

	vector<double> answers{ answer1, answer2, answer3 };
	return answers;

}

vector<double> HaltonSequence(int k, int m)
{
	vector<double> sequence(k, 0);
	double r = ceil(log(k) / log(m)) + 1;
	vector<double> baseRep(r, 0);
	for (int count = 0; count < k; count++)
	{
		int j = 0;
		bool done = false;
		while (!done)
		{
			baseRep[j] = baseRep[j] + 1;
			if (baseRep[j] < m){done = true;}
			else
			{
				baseRep[j] = 0;
				j++;
			}
		}
		for (int baseCt = 0; baseCt < r; baseCt++)
		{
			sequence[count] = sequence[count] + (baseRep[baseCt] / pow(m, baseCt+1));
		}
	}
	return sequence;
}

double CalculateIntegral(int m1, int m2, int k)
{
	vector<double> seq1 = HaltonSequence(k, m1);
	vector<double> seq2 = HaltonSequence(k, m2);

	vector<double> intVal;
	for (int count = 0; count < k; count++)
	{
		double val = exp(-1.0 * seq1[count] * seq2[count])* (sin(6 * M_PI * seq1[count]) + cbrt(cos(2 * M_PI * seq2[count])));
		intVal.push_back(val);
	}
	return mean(intVal);
}